File: cpp/n506_RelativeRank.cpp
// https://leetcode-cn.com/problems/relative-ranks/
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
using namespace std;

class Solution {
public:
    struct Node{
        int index;
        int score;
        Node(int i, int s): index(i), score(s) {};
        bool operator < (const Node& other) const{
            if (score == other.score){
                return index < other.index;
            }
            return score > other.score;
        }
    };
    
    string trans(int i){
        if (i == 0) return "Gold Medal";
        if (i == 1) return "Silver Medal";
        if (i == 2) return "Bronze Medal";
        return to_string(i+1);
    }
    
    vector<string> findRelativeRanks(vector<int>& nums) {
        int l = nums.size();
        vector<Node> nodes;
        for (int i = 0; i < l; ++i){
            nodes.push_back(Node(i, nums[i]));
        }
        sort(nodes.begin(), nodes.end());
        vector<string> res(l);
        for (int j = 0; j < l; ++j){
            res[nodes[j].index] = trans(j);
        }
        return res;
    }
};

int main(){
    return 0;
}

File: python/n506_RelativeRank.py
# https://leetcode-cn.com/problems/relative-ranks/
class Solution:
    def findRelativeRanks(self, nums):
        res = [0] * len(nums)
        for i, j in enumerate(nums):
            nums[i] = (j, i+1)
        nums.sort(key=lambda x: x[0], reverse=True)      
        for k, t in enumerate(nums):
            res[t[1]-1] = str(k + 1)
        for i in range(len(res)):
            if res[i] == "1": res[i] = "Gold Medal"
            if res[i] == "2": res[i] = "Silver Medal"
            if res[i] == "3": res[i] = "Bronze Medal"
        return res
        

if __name__ == "__main__":
    pass